src/seedu/addressbook/logic/Logic.java

```diff
@@ -78,18 +82,25 @@ public CommandResult execute(String userCommandText) throws Exception {
     /**
      * Executes the command, updates storage, and returns the result.
      *
-     * @param command user command
+     * @param command
+     *            user command
      * @return result of the command
-     * @throws Exception if there was any problem during command execution.
+     * @throws Exception
+     *             if there was any problem during command execution.
      */
     private CommandResult execute(Command command) throws Exception {
         command.setData(addressBook, lastShownList);
         CommandResult result = command.execute();
-        storage.save(addressBook);
+        if (command.isMutating()) {
+            storage.save(addressBook);
+        }
```

Comment:
The behavior of the code has been changed. Ensure the header comment and the tests have been updated.